src/nodes.coffee

```diff
@@ -3026,10 +3026,20 @@ exports.For = class For extends While
       begin = new Value new Literal "_begin"
       end = new Value new Literal "_end"
       positive = new Value new Literal "_positive"
+      stepVal = new Value new Literal "_step"
+
       # Calculate stepping
-      stepVal = @step or new Literal 1
-      step = new If positive, new Op("+=", @name, stepVal)
-      step.addElse new Op("-=", @name, stepVal)
+      if @step
+        a_stepVal = new Assign stepVal, @step
+        a_positive = new Assign positive, (new Op ">", stepVal, new Literal 0), null, { icedlocal : true }
+      else
+        is_positive = (new Op ">", end, begin)
+        a_stepVal = new If is_positive, new Assign stepVal, new Literal 1
+        a_stepVal.addElse new Assign stepVal, new Literal -1
+        a_positive = new Assign positive, is_positive, null, { icedlocal : true }
```

Comment:
We need `{ icedlocal : true }` to make sure scoping works appropriately.  Otherwise, it's scoped to the top-level function, meaning nested loops will exhibit incorrect behavior.